include/range/v3/view/split.hpp

```diff
@@ -151,49 +151,56 @@ namespace ranges
                 struct predicate_pred
                 {
                     semiregular_t<Pred> pred_;
+
+                    template<class S, CONCEPT_REQUIRES_(Sentinel<S, iterator_t<Rng>>())>
                     std::pair<bool, iterator_t<Rng>>
-                    operator()(iterator_t<Rng> cur, sentinel_t<Rng> end) const
+                    operator()(iterator_t<Rng> cur, S end) const
                     {
                         auto where = ranges::find_if_not(cur, end, std::ref(pred_));
-                        return std::make_pair(cur != where, where);
+                        return std::pair<bool, iterator_t<Rng>>{cur != where, where};
                     }
                 };
                 template<typename Rng>
                 struct element_pred
                 {
                     range_value_type_t<Rng> val_;
+
+                    template<class S, CONCEPT_REQUIRES_(Sentinel<S, iterator_t<Rng>>())>
                     std::pair<bool, iterator_t<Rng>>
-                    operator()(iterator_t<Rng> cur, sentinel_t<Rng> end) const
+                    operator()(iterator_t<Rng> cur, S end) const
                     {
-                        using P = std::pair<bool, iterator_t<Rng>>;
                         RANGES_EXPECT(cur != end);
-                        return *cur == val_ ? P{true, ranges::next(cur)} : P{false, cur};
+                        bool const match = *cur == val_;
+                        if (match) ++cur;
+                        return std::pair<bool, iterator_t<Rng>>{match, cur};
```

Comment:
```c++
return {0 != advance(cur, *cur == val, unreachable{}), cur};
```

:-)

Reply:
:gag:

Reply:
This is like playing the "C++11 constexpr" game.

Reply:
You're right. I need to move the iterator:

```c++
return {0 != advance(cur, *cur == val, unreachable{}), std::move(cur)};
```

Perfect! :-D

Reply:
Drat! Foiled! Curses! And if it weren't for you meddling kids....

Reply:
![image](https://user-images.githubusercontent.com/456873/36765494-61208d6e-1be6-11e8-8c16-7e5ab73a0084.png)